state/enableha.go

```diff
@@ -153,9 +154,45 @@ func (st *State) EnableHA(
 		err = errors.Annotatef(err, "failed to enable HA with %d controllers", numControllers)
 		return ControllersChanges{}, err
 	}
+	// Failing to open the SSH proxy port shouldn't block enabling HA.
+	// The user can always open the port manually.
+	if err := st.openSSHProxyPort(); err != nil {
```

Comment:
Note, this function is not inside the above transaction because the primitives for opening SSH ports assume we have a `unit` object, which isn't available until after the transaction executes. I think it's probably more work than its worth to write the raw documents for opening the SSH ports in the transaction above.

Reply:
We want to add the ops to open the port in the `enableHAIntentionOps` method. See the code for adding the unit below - we want to add to that. We already have the means to get the ops for opening a port.
```
		if controllerApp != nil {
			_, unitOps, err := controllerApp.addUnitOps("", AddUnitParams{
				UnitName:  &controllerUnitName,
				machineID: mdoc.Id,
			}, nil)
			if err != nil {
				return nil, ControllersChanges{}, errors.Trace(err)
			}
			ops = append(ops, unitOps...)
		}
```

Reply:
I push a change to make this work https://github.com/juju/juju/pull/19180/commits/4ae859af6393b5f9b176669f51522c22e82e9c9c

I tested with enable-ha to new machines - we also need to verify that it works with ha placement where the machine exists and we use --to N to convert that machine to a controller. I suspect the added logic will need to be tweaked slightly since the machine exists but not the unit. So we'll need another pre-check. But you get the idea.